File: packages/scala/oauth/src/main/scala/do/oauth/OAuthClient.scala
package `do`.oauth

import scala.util.{Try, Success, Failure}
import io.circe.Json

/**
 * OAuth device flow client for .do APIs.
 */
class OAuthClient(clientId: String = OAuthClient.DefaultClientId) {
  private val deviceFlow = new DeviceFlow(clientId)
  private val storage = new TokenStorage()

  /**
   * Initiates device authorization flow.
   */
  def authorizeDevice(scope: String = "openid profile email"): Try[DeviceAuthResponse] = {
    deviceFlow.authorize(scope)
  }

  /**
   * Polls for tokens after user authorizes the device.
   */
  def pollForTokens(deviceCode: String, interval: Int, expiresIn: Int): Try[TokenResponse] = {
    deviceFlow.pollForTokens(deviceCode, interval, expiresIn) match {
      case Success(tokens) =>
        storage.saveTokens(tokens)
        Success(tokens)
      case f @ Failure(_) => f
    }
  }

  /**
   * Gets current user info using stored access token.
   */
  def getUser(accessToken: Option[String] = None): Try[Json] = {
    val token = accessToken.orElse(storage.getAccessToken())
    token match {
      case Some(t) => deviceFlow.getUserInfo(t)
      case None => Failure(new Exception("No access token available"))
    }
  }

  /**
   * Interactive login flow.
   */
  def login(scope: String = "openid profile email"): Try[TokenResponse] = {
    authorizeDevice(scope) match {
      case Success(deviceInfo) =>
        println(s"\nTo sign in, visit: ${deviceInfo.verificationUri}")
        println(s"And enter code: ${deviceInfo.userCode}\n")
        pollForTokens(deviceInfo.deviceCode, deviceInfo.interval, deviceInfo.expiresIn)
      case Failure(e) => Failure(e)
    }
  }

  /**
   * Logs out by removing stored tokens.
   */
  def logout(): Try[Unit] = {
    storage.deleteTokens()
  }
}

object OAuthClient {
  val DefaultClientId = "client_01JQYTRXK9ZPD8JPJTKDCRB656"

  def apply(clientId: String = DefaultClientId): OAuthClient = new OAuthClient(clientId)
}

File: packages/php/oauth/src/OAuthClient.php
<?php

declare(strict_types=1);

namespace DotDo\OAuth;

/**
 * OAuth device flow client for .do APIs.
 */
class OAuthClient
{
    private DeviceFlow $deviceFlow;
    private TokenStorage $storage;

    public function __construct(
        string $clientId = DeviceFlow::DEFAULT_CLIENT_ID
    ) {
        $this->deviceFlow = new DeviceFlow($clientId);
        $this->storage = new TokenStorage();
    }

    /**
     * Initiates device authorization flow.
     *
     * @param string $scope OAuth scope
     * @return array Device authorization response
     * @throws \Exception on failure
     */
    public function authorizeDevice(string $scope = 'openid profile email'): array
    {
        return $this->deviceFlow->authorize($scope);
    }

    /**
     * Polls for tokens after user authorizes the device.
     *
     * @param string $deviceCode Device code from authorization
     * @param int $interval Polling interval in seconds
     * @param int $expiresIn Expiration time in seconds
     * @return array Token response
     * @throws \Exception on failure
     */
    public function pollForTokens(string $deviceCode, int $interval, int $expiresIn): array
    {
        $tokens = $this->deviceFlow->pollForTokens($deviceCode, $interval, $expiresIn);
        $this->storage->saveTokens($tokens);
        return $tokens;
    }

    /**
     * Gets current user info using stored or provided access token.
     *
     * @param string|null $accessToken Optional access token
     * @return array User info
     * @throws \Exception on failure
     */
    public function getUser(?string $accessToken = null): array
    {
        $token = $accessToken ?? $this->storage->getAccessToken();

        if ($token === null) {
            throw new \Exception('No access token available');
        }

        return $this->deviceFlow->getUserInfo($token);
    }

    /**
     * Interactive login flow.
     *
     * @param string $scope OAuth scope
     * @return array Token response
     * @throws \Exception on failure
     */
    public function login(string $scope = 'openid profile email'): array
    {
        $deviceInfo = $this->authorizeDevice($scope);

        echo "\nTo sign in, visit: {$deviceInfo['verification_uri']}\n";
        echo "And enter code: {$deviceInfo['user_code']}\n\n";

        return $this->pollForTokens(
            $deviceInfo['device_code'],
            $deviceInfo['interval'] ?? 5,
            $deviceInfo['expires_in'] ?? 900
        );
    }

    /**
     * Logs out by removing stored tokens.
     */
    public function logout(): void
    {
        $this->storage->deleteTokens();
    }

    /**
     * Checks if tokens exist.
     */
    public function hasTokens(): bool
    {
        return $this->storage->hasTokens();
    }
}
